[FLINK-6437][History Server] Move history server configuration to a separate file

yanghua opened the pull request:
## What is the purpose of the change

*This pull request move history server configuration to a separate file*

## Brief change log

  - *Move history server configuration to a separate file*


## Verifying this change

*This change has been manually tested (locally compiled and packaged, started with the History Server server)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

yanghua commented:
cc @StephanEwen  @tillrohrmann

zentol commented:
I don't think we should do this _now_. It seems like there's little to gain at this point, since we would still have to support configuration via `flink-conf.yaml` to not break every existing setup (like this PR does).

Given that there's no overlap between HistoryServer and cluster options there's also little chance of accidentally messing up the configuration for the other one.

Besides that this PR is untested (yet again!).

yanghua commented:
@zentol I have read the comments of this issue (no one explicitly objected), I think if you don't need to do it, then we should clearly mark it on JIRA. Also in the PR description, I explained that it passed my local test. I originally wanted to write a test case. This PR change is mainly for `GlobalConfiguration`. I have already seen it. Flink didn't test `flink-conf.yaml` before, and it doesn't seem to provide the infrastructure for this kind of test. The configuration-related tests are all about creating a Configuration object yourself and then setting values for it.

If Flink's original code already contains tests for this configuration file path, I can write tests.

StephanEwen commented:
@yanghua You are right, this was not marked as "later" or anything in JIRA.
There is never really a time where we don't need to worry about backwards compatibility, so any change that addresses this needs to take this into account.

One way to do the backwards compatibility would be to look for the dedicated `flink-historyserver-conf.yaml` config file and fall back to the `flink-conf.yaml` file if the former does not exist.

zentol commented:
That strategy is already implemented, but it doesn't really address backwards compatibility imo. I would assume that when people upgrade they'll end up with the default `flink-historyserver-conf.yaml` being present in `conf` overwriting everything in `flink-conf.yaml`.

We could comment out everything in in the HS config file, always read both and prioritize contents in the HS config. This wouldn't affect old users (we could also guide them with logging messages if settings are found in `flink-conf.yaml`, nor should it affect new users as either a) they have to set a key anyway or b) a sane default handles this case.

Still, I'm wondering whether there's really a benefit here. If we start splitting config files I'd prefer if we'd to the same for the client.

StephanEwen commented:
Sure, so the compatibility story needs a more detailed design, and that needs to be discussed before looking at concrete code. Agreed.

On whether we want this feature or not - the discussion in Jira was in favor of doing this. Curious what is the reasoning for the push back now, @zentol ?

zentol commented:
well I still like the idea of separate config files, but the JIRA discussion happened
more than a year ago, _before we even had released the HistoryServer_. Now we have to think about backwards compatibility and will thus naturally end up adding complexity. I'm not sure if this is really worth it, especially so since this issue has never been raised again since the HS was released.

yanghua commented:
hi @StephanEwen @zentol ,

Currently, the history server runs as a standalone JVM process, which means it is a separate component. I think the configuration split is reasonable. If I only need to start the history server then I don't rely on other configurations, and I don't need to pay attention to other configurations.

Regarding the discussion of compatibility, I basically agree with @zentol  's point of view.

Basic on the existing implementation, I think we can add more log warnings in the new method of `GlobalConfiguration`. The general idea is as follows:

* Judge the new configuration file, if it does not exist, I will add a log to inform the user that he is using the old configuration;
* If the new configuration and the old configuration about history server exist at the same time,  then we will warn the user that there are two configurations and we will base on the `flink-historyserver-conf.yaml`;

In addition, I will give a log warning when the configuration file is loaded by the `HistoryServer#main` method, for example : 

> the current configuration about history server in `flink-conf.yaml` is only for backward compatibility, it is recommended that they enable the new configuration file `flink-historyserver-conf.yaml`.

In addition, I will comment out the configuration items in the `flink-historyserver-conf.yaml`, considering that we have done so many guides, I believe users will be more likely to accept the correct guidelines.

Even if it is based on a configuration already in `flink-conf.yaml`, the current implementation will still work. But if we say that there are configurations in both files, we seem to have no better way than to give a warning "based on `flink-historyserver-conf.yaml`".

yanghua commented:
@StephanEwen @zentol and @tillrohrmann I have refactored this PR and added test cases, following some of the ideas I stated in the comments. Can you help me with a review? Thank you!

yanghua commented:
Hi @StephanEwen What do you think of this change?

yanghua commented:
@StephanEwen @tillrohrmann and @zentol ,Any opinions about this PR?

zentol commented:
My opinion hasn't changed, I still don't think that it is worth it.

Separating the HistoryServer from your cluster environment is already trivial: just copy flink-dist to separate locations.
The downside here is you end up with a plethora of unused jars in the HistoryServer copy, but I'd argue that the HistoryServer shouldn't be bundled with flink-dist in such a messy way and instead be distributed as a standalone application. However this shouldn't be discussed in this PR.

yanghua commented:
@zentol OK, so what do you think is the correct way to handle it? How to separate History Server? If your solution is more reasonable, then I am happy to accept and implement it.

tillrohrmann commented:
Closing because this PR has been abandoned/not reached consensus.
